### PR TITLE
Option to resume previously interrupted calibration, fix edge cases

### PR DESCRIPTION
**Pull request by justheuristic**
This PR introduces an option to resume calibration from partially quantized model, with the following use cases
- when quantizing a particularly large model (looking at you, grok!), you may run the job every night and stop it during daytime to free GPUs for other tasks
- if the original run crashed due to OOM or other means, one can resume without losing (almost) any progress
- when calibrating on preemptible hardware (e.g. spot instances), one can begin training on one server, then switch to another server when the first one is preempted.
- when calibration went well for most layers, but then loss exploded at *some* layer close to the end (e.g. due to too large an LR), one can change parameters and re-train from that layer on

To this end, the PR introduces two new options to `main.py`:
- `--resume` - if enabled, the calibration code will check if `--save` path already contains previously quantized blocks
   - if a pre-quantized block exists, the code will load it instead of 1) calibrating individual linear layers within that block and 2) fine-tuning the block itself to optimize MSE
  - this option **requires** `--save` to be enabled
- `--on_save="import my_cloud_saver as s; s.save_to_cloud()"` - users may specify the code to run after each block
  - the most obvious use case is when running on preemptible VMs; use this option to upload the quantized blocks to persistent storage (e.g. S3 for AWS spot VMs)
  - another less obvious use case is to upload the model to HF hub as it is quantized
  - since you can write python code here, you can also make it send you notifications or do some other bizarre things

Note that recovering from partial quantization is not instantaneous: the model still needs to compute input activations  by running forward pass on any previously quantized blocks. This takes some time, but is typically 1 or 2 orders of magnitude faster than quantizing them from scratch, depending on the number of training data points. The reason we re-run forward pass and not save these activations is that they are typically huge, much larger than the entire model.

### CODE REVIEW

File: finetune.py [hunk 1]
```diff
@@ -365,15 +365,13 @@ def finetune(model, train_loader, train_hiddens, args, device, val_loader=None,
         shutil.copy(os.path.join(args.quant_model, "args.pt"), os.path.join(args.save, "args.pt"))
 
     print("\n============ Evaluating perplexity... ============")
-    if args.eval_model_seqlen:
-        quant_model.seqlen = args.eval_model_seqlen
```
Comment (justheuristic) [inline finetune.py:369]: Rationale: it is best not to rely on a monkey-patched additional property (.seqlen) **that is not originally a part of model**. It seems relatively easy to pass seqlen directly instead.

File: main.py [hunk 2]
```diff
@@ -34,56 +34,56 @@
...
 @torch.no_grad()
 def get_inps(
-    model: PreTrainedModel, data_iterable: Iterable, args: Namespace, nsamples: Optional[int] = None
-) -> Sequence[torch.Tensor]:
+    model: PreTrainedModel,
+    data: Sequence,
+    model_seqlen: int,
+    devices: Sequence[torch.device],
+    offload_activations: bool,
+) -> Tuple[Sequence[torch.Tensor], Dict]:
```
Comment (justheuristic) [inline main.py:69]: I tested that running perplexity_eval with old vs new get_inps returns exactly the same perplexity

File: main.py [hunk 3]
```diff
@@ -34,56 +34,56 @@
     has_wandb = False
 
 
-def quantize_model(model, args):
+def quantize_model(model: PreTrainedModel, args: Namespace):
     """main entry point to functions for model quantization"""
     tick = time.time()
     print("Loading data ...")
-    dataloader = get_loaders(
+    data = get_loaders(
```
Comment (justheuristic) [inline main.py:41]: Rationale: this is not actually a dataloader, but a list of sequences; the previous name is misleading.

File: main.py [hunk 4]
```diff
@@ -154,34 +155,32 @@ def forward(self, inp, **kwargs):
...
 
-    num_codebooks = args.num_codebooks
     use_cache = model.config.use_cache
     num_codebooks = args.num_codebooks
     model.config.use_cache = False
 
     quantizers = {}
     overall_bits = 0
-    model_number_of_params = 0
+    number_of_quantized_params = 0
```
Comment (justheuristic) [inline main.py:183]: Rationale: this is the number of parameters quantized by this run, which is not the same as all model parameters.

File: main.py [hunk 5]
```diff
@@ -199,12 +198,21 @@ def quantize_aq(
             forward_args[k] = v.to(args.devices[0]) if isinstance(v, torch.Tensor) else v
 
         if args.true_sequential:
-            sequential = get_sequential_groups(model, args.mlp_only)
+            sequential = get_sequential_groups(model)
```
Comment (justheuristic) [inline main.py:201]: This used to crash (no such parameter in args)

File: main.py [hunk 6]
```diff
@@ -214,6 +222,9 @@ def quantize_aq(
                 )
 
         for names in sequential:
+            if loaded_layer:
+                print("Skipping quantization: loaded a previously quantized layer")
+                break
```
Comment (justheuristic) [inline main.py:227]: Why this is inside the for loop: so as not to add another nested IF around the entire thing. This makes black do extreme line breaks.

File: main.py [hunk 7]
```diff
@@ -271,10 +282,13 @@ def quantize_aq(
 
                 weight_avg_bits = quantized_weight.estimate_nbits_per_parameter()
                 overall_bits += int(weight_avg_bits * torch.numel(aq_handlers[sublayer_name].layer.weight.data))
-                model_number_of_params += torch.numel(aq_handlers[sublayer_name].layer.weight.data)
-                print("curent_avg_bits", overall_bits / model_number_of_params)
+                number_of_quantized_params += torch.numel(aq_handlers[sublayer_name].layer.weight.data)
+                print("curent_avg_bits", overall_bits / number_of_quantized_params)
                 quantizers["model.layers.%d.%s" % (layer_index, sublayer_name)] = ()  # to be updated
 
+            del aq_handlers
```
Comment (justheuristic) [inline main.py:289]: Rationale: this saves a bit of memory by deleting XTX after we no longer need them. We can use this to fit ever so slightly larger models.

File: main.py [hunk 8]
```diff
@@ -303,21 +320,23 @@ def quantize_aq(
             out_losses = update_outs_parallel(
                 args.devices, layer, inps, outs, compute_mse=not args.skip_out_loss, **forward_args
             )
+        stats_payload["out_loss"] = torch.mean(torch.Tensor(out_losses)).item()
 
         if run_validation:
+            should_compute_mse = not (args.skip_out_loss or loaded_layer)
```
Comment (justheuristic) [inline main.py:326]: Rationale: if we loaded a layer, we do not compute outs before quantization (so as to save time)

File: main.py [hunk 9]
```diff
@@ -290,34 +304,39 @@ def quantize_aq(
...
+
+        if args.save and not loaded_layer:
             os.makedirs(args.save, exist_ok=True)
             layer_save_path = os.path.join(args.save, f"{layer_index}.pth")
             print(f"Saving layer {layer_index}... to {layer_save_path}")
             torch.save(layer, layer_save_path)
+            if args.on_save:
+                exec(args.on_save)  # a callback e.g. to save progress in slurm or similar distributed infrastructure
 
+        should_compute_mse = not (args.skip_out_loss or loaded_layer)
```
Comment (justheuristic) [inline main.py:316]: Rationale: if we loaded the layer, we do not record pre-quantization outputs to save time; hence, we do not have the references for mse. Doing so would increase the time it takes to resume by at least 2x.

File: finetune.py [hunk 10]
```diff
@@ -322,7 +322,7 @@ def finetune(model, train_loader, train_hiddens, args, device, val_loader=None,
         train_dataloader = dataloader
         val_dataloader = None
     # create original model
-    orig_model = get_model(args.base_model, None, args.dtype, args.model_seqlen, args.device_map)
+    orig_model = get_model(args.base_model, None, args.dtype, args.device_map)
```
Comment (justheuristic) [inline finetune.py:325]: rationale: do not attach seqlen as a monkey-patched field of the model; instead, use args.model_seqlen directly.

File: main.py [hunk 11]
```diff
@@ -326,37 +345,38 @@ def quantize_aq(
...
+        if number_of_quantized_params > 0:  # do not report avg bits if we load all pre-quantized layers via --resume
+            wandb.log({"Avg_bits": overall_bits / number_of_quantized_params})
     model.config.use_cache = use_cache
     print(f"quantize: {torch.cuda.max_memory_allocated()=:,}")
     return quantizers
 
 
 @torch.no_grad()
-def perplexity_eval(model, testenc, args):
+def perplexity_eval(model: PreTrainedModel, testenc: torch.LongTensor, args: Namespace) -> float:
```
Comment (justheuristic) [inline main.py:371]: I tested that running perplexity_eval with old vs new get_inps returns exactly the same perplexity

**Review by galqiwi — APPROVED** [review]
LGTM
